File: microservice-architecture/resilience-patterns/go/src/resilience.go
// Package src 实现了弹性模式组合（Resilience Patterns）的核心逻辑。
//
// 本模块将超时、重试、断路器三种弹性模式组合在一起，展示它们如何协同工作。
//
// 【设计模式】
//   - 策略模式（Strategy Pattern）：超时、重试、断路器是三种可独立使用或组合的弹性策略。
//   - 代理模式（Proxy Pattern）：CallWithTimeout / Retry / CircuitBreaker.Execute
//     包裹在真实操作之外，透明地添加弹性行为。
//   - 状态模式（State Pattern）：CircuitBreaker 在 closed/open 状态下行为不同。
//   - 模板方法模式（Template Method）：Retry 定义了循环调用的固定骨架。
//
// 【架构思想】
//   超时防止无限等待，重试处理暂时性故障，断路器阻止级联雪崩。
//
// 【开源对比】
//   - sony/gobreaker：Go 断路器库
//   - avast/retry-go：Go 重试库，支持指数退避
//   - failsafe-go：Go 弹性库，支持策略组合
//   本示例省略了指数退避、goroutine 安全等工程细节，聚焦于三种模式的核心逻辑。
package src

import (
	"errors"
	"time"
)

var ErrTimeout = errors.New("operation timed out")
var ErrCircuitOpen = errors.New("circuit breaker is open")

type Result struct {
	Value string
	Err   error
	Delay time.Duration
}

type ScriptedDependency struct {
	results []Result
	index   int
}

func NewScriptedDependency(results []Result) *ScriptedDependency {
	return &ScriptedDependency{results: results}
}

func (d *ScriptedDependency) Call() (string, error) {
	if len(d.results) == 0 {
		return "", errors.New("no scripted result available")
	}

	index := d.index
	if index >= len(d.results) {
		index = len(d.results) - 1
	}
	result := d.results[index]
	d.index++

	if result.Delay > 0 {
		time.Sleep(result.Delay)
	}

	return result.Value, result.Err
}

func CallWithTimeout(timeout time.Duration, operation func() (string, error)) (string, error) {
	type response struct {
		value string
		err   error
	}

	done := make(chan response, 1)
	go func() {
		value, err := operation()
		done <- response{value: value, err: err}
	}()

	select {
	case result := <-done:
		return result.value, result.err
	case <-time.After(timeout):
		return "", ErrTimeout
	}
}

func Retry(maxAttempts int, operation func() (string, error)) (string, int, error) {
	if maxAttempts <= 0 {
		maxAttempts = 1
	}

	var lastErr error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		value, err := operation()
		if err == nil {
			return value, attempt, nil
		}
		lastErr = err
	}

	return "", maxAttempts, lastErr
}

type CircuitBreaker struct {
	failureThreshold    int
	consecutiveFailures int
	open                bool
}

func NewCircuitBreaker(failureThreshold int) *CircuitBreaker {
	if failureThreshold <= 0 {
		failureThreshold = 1
	}
	return &CircuitBreaker{failureThreshold: failureThreshold}
}

func (c *CircuitBreaker) Execute(operation func() (string, error), fallback string) (string, error) {
	if c.open {
		return fallback, ErrCircuitOpen
	}

	value, err := operation()
	if err != nil {
		c.consecutiveFailures++
		if c.consecutiveFailures >= c.failureThreshold {
			c.open = true
		}
		return fallback, err
	}

	c.consecutiveFailures = 0
	return value, nil
}

func (c *CircuitBreaker) Reset() {
	c.open = false
	c.consecutiveFailures = 0
}

File: microservice-architecture/resilience-patterns/c/src/resilience.c
#include "func.h"

#include <stddef.h>

void scripted_dependency_init(ScriptedDependency *dependency, const ScriptedResult *results, int count)
{
    dependency->count = count;
    dependency->index = 0;
    for (int i = 0; i < count; i++) {
        dependency->results[i] = results[i];
    }
}

static int scripted_call(ScriptedDependency *dependency, const char **value_out, int *delay_ms_out)
{
    int index = dependency->index < dependency->count ? dependency->index : dependency->count - 1;
    ScriptedResult result = dependency->results[index];
    dependency->index++;
    *delay_ms_out = result.delay_ms;
    *value_out = result.value;
    return result.fail ? 0 : 1;
}

int call_with_timeout(ScriptedDependency *dependency, int timeout_ms, const char **value_out)
{
    int delay_ms = 0;
    const char *value = NULL;
    int ok = scripted_call(dependency, &value, &delay_ms);
    if (delay_ms > timeout_ms) {
        return 0;
    }
    if (!ok) {
        return 0;
    }
    *value_out = value;
    return 1;
}

int retry_call(ScriptedDependency *dependency, int max_attempts, const char **value_out, int *attempts_out)
{
    for (int attempt = 1; attempt <= max_attempts; attempt++) {
        int delay_ms = 0;
        const char *value = NULL;
        if (scripted_call(dependency, &value, &delay_ms)) {
            *value_out = value;
            *attempts_out = attempt;
            return 1;
        }
    }
    *attempts_out = max_attempts;
    return 0;
}

void circuit_breaker_init(CircuitBreaker *breaker, int failure_threshold)
{
    breaker->failure_threshold = failure_threshold > 0 ? failure_threshold : 1;
    breaker->consecutive_failures = 0;
    breaker->open = 0;
}

int circuit_breaker_execute(CircuitBreaker *breaker, ScriptedDependency *dependency, const char *fallback, const char **value_out, int *circuit_open)
{
    if (breaker->open) {
        *value_out = fallback;
        *circuit_open = 1;
        return 0;
    }

    int delay_ms = 0;
    const char *value = NULL;
    if (scripted_call(dependency, &value, &delay_ms)) {
        breaker->consecutive_failures = 0;
        *value_out = value;
        *circuit_open = 0;
        return 1;
    }

    breaker->consecutive_failures++;
    if (breaker->consecutive_failures >= breaker->failure_threshold) {
        breaker->open = 1;
    }
    *value_out = fallback;
    *circuit_open = 0;
    return 0;
}

void circuit_breaker_reset(CircuitBreaker *breaker)
{
    breaker->consecutive_failures = 0;
    breaker->open = 0;
}
